File: adapters/src/main/java/com/wang/container/interfaces/IListAdapter.java
package com.wang.container.interfaces;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.ViewDataBinding;

import com.wang.container.holder.BaseViewHolder;

import java.util.Collection;
import java.util.List;

/**
 * 所有list的adapter的接口
 */
public interface IListAdapter<BEANS, DB extends ViewDataBinding, LISTENER extends IItemClick> extends IAdapter<LISTENER> {

    default void setHeaderView(@NonNull Context context, @LayoutRes int layoutRes) {
        if (layoutRes == 0) {
            setHeaderView(null);
        } else {
            setHeaderView(LayoutInflater.from(context).inflate(layoutRes, new FrameLayout(context), false));
        }
    }

    /**
     * @param view null表示删除，view的parent为FrameLayout，默认match、wrap
     */
    void setHeaderView(@Nullable View view);

    @Nullable
    View getHeaderView();

    default void setFooterView(@NonNull Context context, @LayoutRes int layoutRes) {
        if (layoutRes == 0) {
            setFooterView(null);
        } else {
            setFooterView(LayoutInflater.from(context).inflate(layoutRes, new FrameLayout(context), false));
        }
    }

    /**
     * @param view null表示删除，view的parent为FrameLayout，默认match、wrap
     */
    void setFooterView(@Nullable View view);

    @Nullable
    View getFooterView();

    /**
     * 最终你的list的create
     * <p>
     * 默认用DataBinding create
     * 完全不需要的话覆盖整个方法就行了，不会出问题
     * 你也可以重写来添加自己的默认逻辑，如：全局隐藏显示、嵌套rv的默认属性设置等
     */
    @NonNull
    BaseViewHolder<DB> onCreateListViewHolder(@NonNull ViewGroup parent);


    /**
     * 最终你的list的bind
     * <p>
     * 你只需要处理其他数据、其他点击事件等，基本的都加上了：
     * <p>
     * 不需要绑定bean数据，已经默认绑定
     * 不需要调用{@link ViewDataBinding#executePendingBindings()}，已经默认加上
     *
     * @param listPosition 已经做过处理,就是list的position
     */
    void onBindListViewHolder(@NonNull BaseViewHolder<DB> holder, int listPosition, BEANS bean);


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // list相关的方法
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @NonNull
    List<BEANS> getList();

    default void setListAndNotifyDataSetChanged(@Nullable List<? extends BEANS> list) {
        if (list != getList()) {//同一个对象当然啥都不需要干了
            getList().clear();
            if (list != null) {
                getList().addAll(list);
            }
        }
        notifyDataSetChanged();
    }

    /**
     * 获取指定bean
     *
     * @throws IndexOutOfBoundsException 不用多说吧
     */
    @NonNull
    default BEANS get(int listPosition) {
        return getList().get(listPosition);
    }

    /**
     * 清空list,不刷新adapter
     */
    default void clear() {
        getList().clear();
    }

    /**
     * 添加全部条目,不刷新adapter
     */
    default void addAll(@Nullable Collection<? extends BEANS> addList) {
        if (addList != null && getList() != addList) {
            getList().addAll(addList);
        }
    }

    default int size() {
        return getList().size();
    }

    /**
     * list是否为空
     */
    default boolean isEmptyList() {
        return getList().isEmpty();
    }
}

File: adapters/src/main/java/com/wang/container/interfaces/IListAdapter.kt
package com.wang.container.interfaces

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.LayoutRes

/**
 * 所有list的adapter的接口
 */
interface IListAdapter<BEAN> : IAdapter {

    /**
     * 支持手动设置FrameLayout.LayoutParams的属性：with、height、margin，不支持gravity
     * 如果没有Params则默认宽高为match、wrap
     */
    var headerView: View?
    var footerView: View?

    val headerViewCount get() = if (headerView == null) 0 else 1
    val isHeaderView get() = headerView != null
    val footerViewCount get() = if (footerView == null) 0 else 1
    val isFooterView get() = footerView != null

    fun removeHeaderView() {
        headerView = null
    }

    fun removeFooterView() {
        footerView = null
    }

    fun setHeaderView(context: Context, @LayoutRes layoutRes: Int) {
        if (layoutRes == 0) {
            removeHeaderView()
            return
        }
        headerView =
            LayoutInflater.from(context).inflate(layoutRes, FrameLayout(context), false)
    }

    fun setFooterView(context: Context, @LayoutRes layoutRes: Int) {
        if (layoutRes == 0) {
            removeFooterView()
            return
        }
        footerView =
            LayoutInflater.from(context).inflate(layoutRes, FrameLayout(context), false)
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // list相关的方法
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    val list: MutableList<BEAN>

    fun setListAndNotify(dataList: List<BEAN>?) {
        if (dataList !== list) { //同一个对象当然啥都不需要干了
            list.clear()
            if (dataList != null) {
                list.addAll(dataList)
            }
        }
        notifyDataSetChanged()
    }

    fun addAllListAndNotify(l: List<BEAN>?) {
        if (!l.isNullOrEmpty()) {
            list.addAll(l)
            notifyListItemRangeInserted(list.size, l.size)
        }
    }

    fun addDataAndNotify(bean: BEAN?) {
        if (bean != null) {
            notifyListItemInserted(list.size, bean)
        }
    }

    fun addFirstListAndNotify(l: List<BEAN>?) {
        if (!l.isNullOrEmpty()) {
            if (l.size == 1) {
                list.add(0, l[0])
            } else {
                val temp = l + list
                list.clear()
                list.addAll(temp)
            }
            notifyListItemRangeInserted(0, l.size)
        }
    }

    fun addFirstDataAndNotify(bean: BEAN?) {
        if (bean != null) {
            notifyListItemInserted(0, bean)
        }
    }

    /**
     * 获取指定bean
     *
     * @throws IndexOutOfBoundsException 不用多说吧
     */
    fun getItemData(listPosition: Int): BEAN {
        return list[listPosition]
    }

    fun getItemDataOrNull(listPosition: Int): BEAN? {
        return list.getOrNull(listPosition)
    }

    /**
     * 清空list,不刷新adapter
     */
    fun clearList() {
        list.clear()
    }

    /**
     * 添加全部条目,不刷新adapter，[addAllListAndNotify]
     */
    fun addAllList(addList: Collection<BEAN>?) {
        if (addList != null && list !== addList) {
            list.addAll(addList)
        }
    }

    fun listSize(): Int {
        return list.size
    }

    /**
     * list是否为空
     */
    fun isEmptyList(): Boolean {
        return list.isEmpty()
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////
    // notify相关方法
    /////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * 刷新list的position，解决[notifyItemChanged]的position问题
     * @param newBean 新数据
     *                null：你已经自己更新过了，这里只需要调用更新数据
     *                notnull：更新该条数据
     */
    fun notifyListItemChanged(listPosition: Int, newBean: BEAN? = null) {
        if (listPosition < 0 || listPosition >= list.size) {
            return
        }
        if (newBean != null) {
            list[listPosition] = newBean
        }
        notifyItemChanged(listPosition + headerViewCount)
    }

    fun notifyListItemChanged(bean: BEAN) {
        notifyListItemChanged(list.indexOf(bean))
    }

    fun notifyListItemRangeChanged(listPositionStart: Int, itemCount: Int) {
        notifyItemRangeChanged(listPositionStart + headerViewCount, itemCount)
    }

    /**
     * @param insertBean 要插入的数据
     *                   null：你已经自己插入数据了，这里只需要调用更新数据
     *                   notnull：插入该条数据
     */
    fun notifyListItemInserted(listPosition: Int, insertBean: BEAN? = null) {
        if (listPosition < 0) {
            return
        }
        if (insertBean != null) {
            if (listPosition > list.size) {
                return
            }
            list.add(listPosition, insertBean)
        }
        notifyItemInserted(listPosition + headerViewCount)
    }

    fun notifyListItemRangeInserted(listPositionStart: Int, itemCount: Int) {
        notifyItemRangeInserted(listPositionStart + headerViewCount, itemCount)
    }

    fun notifyListItemMoved(listFromPosition: Int, listToPosition: Int) {
        notifyItemMoved(listFromPosition + headerViewCount, listToPosition + headerViewCount)
    }

    /**
     * @param isRemoData 是否删除该数据
     *                false：你已经自己删除过了，这里只需要调用更新数据
     *                true：删除该条数据
     */
    fun notifyListItemRemoved(listPosition: Int, isRemoData: Boolean = false) {
        if (listPosition < 0) {
            return
        }
        if (isRemoData) {
            if (listPosition >= list.size) {
                return
            }
            list.removeAt(listPosition)
        }
        notifyItemRemoved(listPosition + headerViewCount)
    }

    fun notifyListItemRangeRemoved(listPositionStart: Int, itemCount: Int) {
        notifyItemRangeRemoved(listPositionStart + headerViewCount, itemCount)
    }
}